### ISSUE TITLE
Unexpected miner crash

### ISSUE THREAD

**Issue by ribasushi** [filecoin-project/lotus#2988]
**Describe the bug**
Miner crashed after ~3 days of operation

**Version (run `lotus version`):**
[`0d1174cb (origin/next)`](https://github.com/filecoin-project/lotus/tree/0d1174cb) + calibration genesis

**Additional context**
```
...
2020-08-11T16:09:38.346Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:38.472Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:38.488Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:38.505Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:38.555Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:38.572Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:38.590Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:38.605Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:38.622Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:38.639Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:38.656Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:38.739Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:38.772Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:38.778Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:39.178Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:39.273Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:39.415Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:39.481Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:39.498Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:39.515Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:39.531Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:39.857Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.057Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.157Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.286Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.635Z        INFO    messagepool     messagepool/selection.go:258    get pending messages done       {"took": 0.000067021}
2020-08-11T16:09:40.646Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.694Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.713Z        WARN    messagepool     messagepool/selection.go:411    encountered message from actor t3tg5il5rmcymtytue7vue2ibscc2kng6a3kprefcqluayg2uuvjhxuasmenps5k7mq2r7dfjdz53kbyq5awga with nonce (15379) less than the current nonce (15467)
2020-08-11T16:09:40.719Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.736Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.771Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.786Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.803Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.828Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.845Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.861Z        INFO    pubsub  go-libp2p-pubsub@v0.3.4-0.20200731161531-2b5243c72f0d/pubsub.go:927     received message from router graylisted peer 12D3KooWLKvqkJe5uPTVhZQGoegjCdabVzXcomXnwr7QJ2D84AaZ. Dropping RPC
2020-08-11T16:09:40.870Z        INFO    messagepool     messagepool/pruning.go:42       message pruning took 236.157875ms
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x115242a]

goroutine 706 [running]:
github.com/filecoin-project/lotus/chain/messagepool.(*msgChain).Before(...)
        /srv/lotus/lotus/chain/messagepool/selection.go:538
github.com/filecoin-project/lotus/chain/messagepool.(*MessagePool).pruneMessages.func2(0x0, 0x23, 0x0)
        /srv/lotus/lotus/chain/messagepool/pruning.go:81 +0x4a
sort.doPivot_func(0xc00ba21800, 0xc042bef7a0, 0x0, 0x84, 0x18, 0xc042bef780)
        /usr/local/go/src/sort/zfuncversion.go:87 +0x44c
sort.quickSort_func(0xc00ba21800, 0xc042bef7a0, 0x0, 0x84, 0x10)
        /usr/local/go/src/sort/zfuncversion.go:143 +0x9a
sort.Slice(0x25bcf60, 0xc042bef780, 0xc00ba21800)
        /usr/local/go/src/sort/slice.go:17 +0xf1
github.com/filecoin-project/lotus/chain/messagepool.(*MessagePool).pruneMessages(0xc01e944f70, 0x2d7c180, 0xc00003e0c0, 0xc05173f400, 0x0, 0x0)
        /srv/lotus/lotus/chain/messagepool/pruning.go:80 +0x6c7
github.com/filecoin-project/lotus/chain/messagepool.(*MessagePool).pruneExcessMessages(0xc01e944f70, 0x0, 0x0)
        /srv/lotus/lotus/chain/messagepool/pruning.go:28 +0x11b
github.com/filecoin-project/lotus/chain/messagepool.(*MessagePool).runLoop(0xc01e944f70)
        /srv/lotus/lotus/chain/messagepool/messagepool.go:342 +0xa38
created by github.com/filecoin-project/lotus/chain/messagepool.New
        /srv/lotus/lotus/chain/messagepool/messagepool.go:249 +0x638
```

**Comment by vyzo** [comment]
The bug has been fixed in https://github.com/filecoin-project/lotus/pull/2979 which will merge soon.

**Comment by RobQuistNL** [comment]
Just for logging purposes..

Also crashed :)

```
2020-08-11T18:18:57.959Z	WARN	messagepool	messagepool/selection.go:395	encountered message from actor t3tg5il5rmcymtytue7vue2ibscc2kng6a3kprefcqluayg2uuvjhxuasmenps5k7mq2r7dfjdz53kbyq5awga with nonce (15379) less than the current nonce (15467)
2020-08-11T18:18:58.073Z	INFO	messagepool	messagepool/pruning.go:42	message pruning took 154.895235ms
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x1151d4a]

goroutine 741 [running]:
github.com/filecoin-project/lotus/chain/messagepool.(*msgChain).Before(...)
	/home/filecoin/lotus/chain/messagepool/selection.go:522
github.com/filecoin-project/lotus/chain/messagepool.(*MessagePool).pruneMessages.func2(0x0, 0x49, 0x0)
	/home/filecoin/lotus/chain/messagepool/pruning.go:81 +0x4a
sort.doPivot_func(0xc04916f800, 0xc08e1f0e60, 0x0, 0x102, 0x18, 0xc08e1f0e40)
	/usr/local/go/src/sort/zfuncversion.go:87 +0x44c
sort.quickSort_func(0xc04916f800, 0xc08e1f0e60, 0x0, 0x102, 0x12)
	/usr/local/go/src/sort/zfuncversion.go:143 +0x9a
sort.Slice(0x25bbf60, 0xc08e1f0e40, 0xc04916f800)
	/usr/local/go/src/sort/slice.go:17 +0xf1
github.com/filecoin-project/lotus/chain/messagepool.(*MessagePool).pruneMessages(0xc000ce3040, 0x2d7aec0, 0xc00003e0b8, 0xc01584e240, 0x0, 0x0)
	/home/filecoin/lotus/chain/messagepool/pruning.go:80 +0x6c7
github.com/filecoin-project/lotus/chain/messagepool.(*MessagePool).pruneExcessMessages(0xc000ce3040, 0x0, 0x0)
	/home/filecoin/lotus/chain/messagepool/pruning.go:28 +0x11b
github.com/filecoin-project/lotus/chain/messagepool.(*MessagePool).runLoop(0xc000ce3040)
	/home/filecoin/lotus/chain/messagepool/messagepool.go:342 +0xa38
created by github.com/filecoin-project/lotus/chain/messagepool.New
	/home/filecoin/lotus/chain/messagepool/messagepool.go:249 +0x638

```

**Comment by arajasek** [comment]
Fixed in #2979
